(wip) add init container that waits for apiext

AliceProxy opened the pull request:
## Description

Adds an init container to the Edge Stack deployment that will wait for apiext to be available and ready before starting up. It also has logic to make sure that apiext is not in the middle of a restart/update as well just in-case. I decided to make this out-out, but we can definitely reverse that choice easily if anyone prefers it that way.

The init container does "log" what it is doing using `echo` just in-case someone is trying to use it and it is not working properly. This would probably only be caused by something like modifying the installation location of apiext and not configuring the helm values properly, but the defaults should be correct for everyone otherwise. Hopefully this fixes some flakes in CI as well since I've seen a bunch of failures lately that appear to be caused by apiext.

(whoops this was supposed to be a draft but I forgot)

Notes from Tenshin:
Unfortunately I couldn't get the tests passing with the actual init container turned on. It doesn't seem to have anything to do with the functionality of the container and more so how we are generating the config for the tests. After testing, I believe this PR is safe to merge despite the functionality not being used in the tests. However, due to this, I changed a few things, namely making it opt out rather than opt in. I also changed the security context to not run as root by default and added the option to change this in the values.yaml.  

## Sample output

### emissary-apiext already exists and is ready
```
checking if deployment/apiext deployment in namespace: emissary-system exists.
emissary-apiext.emissary-system exists.
checking if emissary-apiext.emissary-system is fully available...
deployment.apps/emissary-apiext condition met
emissary-apiext.emissary-system is available
emissary-apiext.emissary-system is fully ready and not currently restarting.  Have: 3, want 3
```

### emissary-apiext already exists, but is in the process of restarting
```
checking if deployment/emissary-apiext in namespace: emissary-system exists.
emissary-apiext.emissary-system exists.
checking if emissary-apiext.emissary-system is fully available...
deployment.apps/emissary-apiext condition met
emissary-apiext.emissary-system is available
emissary-apiext.emissary-system is in the process of restarting. Have: 4, want 3
emissary-apiext.emissary-system is fully ready and not currently restarting.  Have: 3, want 3
```

# emissary-apiext does not exist yet
```
checking if deployment/emissary-apiext in namespace: emissary-system exists.
emissary-apiext.emissary-system does not exist yet. Waiting...
checking if deployment/emissary-apiext in namespace: emissary-system exists.
emissary-apiext.emissary-system does not exist yet. Waiting...
checking if deployment/emissary-apiext in namespace: emissary-system exists.
emissary-apiext.emissary-system does not exist yet. Waiting...
emissary-apiext.emissary-system exists.
checking if emissary-apiext.emissary-system is fully available...
deployment.apps/emissary-apiext condition met
emissary-apiext.emissary-system is available
emissary-apiext.emissary-system is fully ready and not currently restarting.  Have: 3, want 3
```

## Related Issues

List related issues.

## Testing

A few sentences describing what testing you've done, e.g., manual tests, automated tests, deployed in production, etc.

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->
- [ ] **Does my change need to be backported to a previous release?**
  - What backport versions were discussed with the Maintainers in the Issue?

- [ ] **I made sure to update `CHANGELOG.md`.**

   Remember, the CHANGELOG needs to mention:
  - Any new features
  - Any changes to our included version of Envoy
  - Any non-backward-compatible changes
  - Any deprecations

- [ ] **This is unlikely to impact how Ambassador performs at scale.**

   Remember, things that might have an impact at scale include:
  - Any significant changes in memory use that might require adjusting the memory limits
  - Any significant changes in CPU use that might require adjusting the CPU limits
  - Anything that might change how many replicas users should use
  - Changes that impact data-plane latency/scalability

- [ ] **My change is adequately tested.**

   Remember when considering testing:
  - Your change needs to be specifically covered by tests.
    - Tests need to cover all the states where your change is relevant: for example, if you add a behavior that can be enabled or disabled, you'll need tests that cover the enabled case and tests that cover the disabled case. It's not sufficient just to test with the behavior enabled.
  - You also need to make sure that the _entire area being changed_ has adequate test coverage.
    - If existing tests don't actually cover the entire area being changed, add tests.
    - This applies even for aspects of the area that you're not changing – check the test coverage, and improve it if needed!
  - We should lean on the bulk of code being covered by unit tests, but...
  - ... an end-to-end test should cover the integration points

- [ ] **I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.**

- [ ] **The changes in this PR have been reviewed for security concerns and adherence to security best practices.**

tenshinhigashi commented:
Unfortunately I couldn't get the tests passing with the actual init container turned on. It doesn't seem to have anything to do with the functionality of the container and more so how we are generating the config for the tests. After testing, I believe this PR is safe to merge despite the functionality not being used in the tests. However, due to this, I changed a few things, namely making it opt out rather than opt in. I also changed the security context to not run as root by default and added the option to change this in the values.yaml.

LanceEa commented:
>Unfortunately I couldn't get the tests passing with the actual init container turned on

I can help you look at this today. Just reviewing the code, I believe the behavior change will help mitigate some of the issues we see with helm installs and gitops tooling. But, I don't think we should rush this in before we make sure it is properly tested and there are no gotchas. 

Have we pulled this into Edge Stack and tested as well?

> However, due to this, I changed a few things, namely making it opt out rather than opt in

I'm guessing the reason for this is to avoid the failing tests or was there a different reason for it? If it was failing test then I would like us to address the test and keep it opt-out instead.

> I also changed the security context to not run as root by default and added the option to change this in the values.yaml.

Since the test are turned off, did we verify there are not issues with running it with the 8888 user?

AliceProxy commented:
> Unfortunately I couldn't get the tests passing with the actual init container turned on. It doesn't seem to have anything to do with the functionality of the container and more so how we are generating the config for the tests. After testing, I believe this PR is safe to merge despite the functionality not being used in the tests. However, due to this, I changed a few things, namely making it opt out rather than opt in. I also changed the security context to not run as root by default and added the option to change this in the values.yaml.

I'm a strong -1 on just disabling test coverage for this and merging it as-is.

AliceProxy reviewed (commented):
I can't actually approve this since I started the PR, but your changes to get it working with the tests seem fine to me although I'd prefer if you squashed all the commits down into one on top of my initial commit, but that is more of a nit so long as you merge by squashing and not rebasing.

LanceEa commented on charts/emissary-ingress/values.yaml.in:
```diff
@@ -254,6 +254,14 @@ restartPolicy:
 terminationGracePeriodSeconds:
 
+waitForApiext:
+  enabled: false
```
I think we want this enabled by default right? Then the user can opt out if they do not want the init containers.

LanceEa replied:
@AliceProxy - thoughts?

LanceEa reviewed (commented):
overall looks good but just had one open question on the defaults for helm.

LanceEa reviewed (approved):
LGTM...when merging, be sure to squash the commits and cleanup the notes.
